# UFO's and Preprocessing
- correcting data types
- removing missing values

In [1]:
import pandas as pd
ufo = pd.read_csv('datasets/ufo_sightings_large.csv')
ufo.head()

date              city state  ...    recorded         lat        long
0   11/3/2011 19:21         woodville    wi  ...  12/12/2011  44.9530556  -92.291111
1   10/3/2004 19:05         cleveland    oh  ...  10/27/2004  41.4994444  -81.695556
2   9/25/2009 21:00       coon rapids    mn  ...  12/12/2009  45.1200000  -93.287500
3  11/21/2002 05:45          clemmons    nc  ...  12/23/2002  36.0213889  -80.382222
4   8/19/2010 12:55  calgary (canada)    ab  ...   8/24/2010   51.083333 -114.083333

[5 rows x 11 columns]

In [2]:
# Print the DataFrame info
print(ufo.info())

# Change the type of seconds to float
ufo["seconds"] = ufo['seconds'].astype(float)

# Change the date column to type datetime
ufo["date"] = pd.to_datetime(ufo['date'])

# Check the column types
print(ufo.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4935 entries, 0 to 4934
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   date            4935 non-null   object 
 1   city            4926 non-null   object 
 2   state           4516 non-null   object 
 3   country         4255 non-null   object 
 4   type            4776 non-null   object 
 5   seconds         4935 non-null   float64
 6   length_of_time  4792 non-null   object 
 7   desc            4932 non-null   object 
 8   recorded        4935 non-null   object 
 9   lat             4935 non-null   object 
 10  long            4935 non-null   float64
dtypes: float64(2), object(9)
memory usage: 424.2+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4935 entries, 0 to 4934
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   date            4935 non-null   d

In [3]:
# Count the missing values in the length_of_time, state, and type columns, in that order
print(ufo[['length_of_time','state', 'type']].isna().sum())

# Drop rows where length_of_time, state, or type are missing
ufo.dropna(subset=['state', 'type', 'length_of_time'],inplace=True)

# Print out the shape of the new dataset
print(ufo.shape)

length_of_time    143
state             419
type              159
dtype: int64
(4283, 11)


# Standardization

In [4]:
import numpy as np

# Check the variance for suitable candidate for standardization
print(ufo.var())

# Log normalize the seconds column
ufo["seconds_log"] = np.log(ufo['seconds'])

# Print out the variance of just the seconds_log column
print(ufo['seconds_log'].var())

seconds    1.545212e+10
long       4.119751e+02
dtype: float64
nan


# Engineering new features
- Extracting numbers from strings
- Encoding categorical variables
- Features from dates
- Text vectorization

**Extracting numbers from strings**

In [5]:
import re

def return_minutes(time_string):

    # Search for numbers in time_string
    num = re.search('\d+', time_string)
    if num is not None:
        return int(num.group(0))
        
# Apply the extraction to the length_of_time column
ufo["minutes"] = ufo["length_of_time"].apply(return_minutes)

# Take a look at the head of both of the columns
print(ufo[['length_of_time','minutes']].head())

    length_of_time  minutes
0          2 weeks      2.0
1           30sec.     30.0
3  about 5 minutes      5.0
4                2      2.0
5       10 minutes     10.0


**Encoding categorical variables**

In [6]:
# Use pandas to encode us values as 1 and others as 0
ufo["country_enc"] = ufo["country"].apply(lambda value: 1 if value == "us" else 0)

# Print the number of unique type values
print(len(ufo['type'].unique()))

# Create a one-hot encoded set of the type values
type_set = pd.get_dummies(ufo['type'])

# Concatenate this set back to the ufo DataFrame
ufo = pd.concat([ufo, type_set], axis=1)

21


In [7]:
ufo

date              city state  ... teardrop triangle  unknown
0    2011-11-03 19:21:00         woodville    wi  ...        0        0        1
1    2004-10-03 19:05:00         cleveland    oh  ...        0        0        0
3    2002-11-21 05:45:00          clemmons    nc  ...        0        1        0
4    2010-08-19 12:55:00  calgary (canada)    ab  ...        0        0        0
5    2012-06-16 23:00:00         san diego    ca  ...        0        0        0
...                  ...               ...   ...  ...      ...      ...      ...
4930 2000-07-05 19:30:00     schnecksville    pa  ...        0        0        0
4931 2008-03-18 22:00:00            gibson    ga  ...        0        1        0
4932 2005-06-15 02:30:00              kent    wa  ...        0        0        0
4933 1991-11-01 03:00:00             niles    mi  ...        0        1        0
4934 2005-12-10 18:00:00           phoenix    az  ...        0        0        0

[4283 rows x 35 columns]

**Features from dates**

In [8]:
# Look at the first 5 rows of the date column
print(ufo['date'].head())

# Extract the month from the date column
ufo["month"] = ufo["date"].dt.month

# Extract the year from the date column
ufo["year"] = ufo["date"].dt.year

# Take a look at the head of all three columns
print(ufo[['date','month','year']].head())

0   2011-11-03 19:21:00
1   2004-10-03 19:05:00
3   2002-11-21 05:45:00
4   2010-08-19 12:55:00
5   2012-06-16 23:00:00
Name: date, dtype: datetime64[ns]
                 date  month  year
0 2011-11-03 19:21:00     11  2011
1 2004-10-03 19:05:00     10  2004
3 2002-11-21 05:45:00     11  2002
4 2010-08-19 12:55:00      8  2010
5 2012-06-16 23:00:00      6  2012


**Text Vectorization**

In [9]:
from sklearn.feature_extraction.text import TfidfVectorizer

# Take a look at the head of the desc field
print(ufo['desc'].head())

# Instantiate the tfidf vectorizer object
vec = TfidfVectorizer()

# Fit and transform desc using vec
desc_tfidf = vec.fit_transform(ufo['desc'])

# Look at the number of columns and rows
print(desc_tfidf.shape)

0    Red blinking objects similar to airplanes or s...
1                 Many fighter jets flying towards UFO
3    It was a large&#44 triangular shaped flying ob...
4       A white spinning disc in the shape of an oval.
5    Dancing lights that would fly around and then ...
Name: desc, dtype: object
(4283, 5754)


# Feature selection and modeling

**Selecting the ideal dataset**

Now to get rid of some of the unnecessary features in the ufo dataset. Because the `country` column has been encoded as `country_enc`, you can select it and drop the other columns related to location: `city`, `country`, `lat`, `long`, and `state`.

You've engineered the `month` and `year` columns, so you no longer need the `date` or `recorded` columns. You also standardized the `seconds` column as `seconds_log`, so you can drop `seconds` and `minutes`.

You vectorized `desc`, so it can be removed. For now you'll keep `type`.

You can also get rid of the `length_of_time` column, which is unnecessary after extracting `minutes`.

In [11]:
# Make a list of features to drop
to_drop = ['city', 'country', 'lat', 'long', 'state','date','recorded','seconds','minutes','desc','length_of_time']

# Drop those features
ufo_dropped = ufo.drop(to_drop,axis=1)

**Filtering words**

In [15]:
# Reverse the vocabulary dictionary 
vocab = {v:k for k,v in vec.vocabulary_.items()}

In [24]:
# Add in the rest of the arguments
def return_weights(vocab, original_vocab, vector, vector_index, top_n):
    zipped = dict(zip(vector[vector_index].indices, vector[vector_index].data))
    
    # Transform that zipped dict into a series
    zipped_series = pd.Series({vocab[i]:zipped[i] for i in vector[vector_index].indices})
    
    # Sort the series to pull out the top n weighted words
    zipped_index = zipped_series.sort_values(ascending=False)[:top_n].index
    return [original_vocab[i] for i in zipped_index]

In [25]:
def words_to_filter(vocab, original_vocab, vector, top_n):
    filter_list = []

    # for all documents
    for i in range(0, vector.shape[0]):
    
        # Call the return_weights function and extend filter_list
        filtered = return_weights(vocab, original_vocab, vector, i, top_n)
        filter_list.extend(filtered)
        
    # Return the list in a set, so we don't get duplicate word indices
    return set(filter_list)

In [26]:
# Let's also filter some words out of the text vector we created
filtered_words = words_to_filter(vocab, vec.vocabulary_, desc_tfidf, 4)

In [28]:
print(filtered_words)

{0, 1, 2, 3, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 180, 181, 182, 183, 184, 185, 186, 187, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 202, 203, 204, 205, 206, 208, 209, 210, 211, 212, 213, 214, 215, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 

In [30]:
len(filtered_words)

5613

In [31]:
desc_tfidf.shape

(4283, 5754)

Words were filtered from 5754 to 5613.

**Modeling the UFO dataset, part 1**

In [35]:
X = ufo_dropped.drop(['country_enc','type'],axis=1)
y = ufo_dropped['country_enc']

In [36]:
# Take a look at the features in the X set of data
print(X.columns)

Index(['seconds_log', 'changing', 'chevron', 'cigar', 'circle', 'cone',
       'cross', 'cylinder', 'diamond', 'disk', 'egg', 'fireball', 'flash',
       'formation', 'light', 'other', 'oval', 'rectangle', 'sphere',
       'teardrop', 'triangle', 'unknown', 'month', 'year'],
      dtype='object')


In [41]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.impute import SimpleImputer
import numpy as np

# Replace any infinite values in X with NaN
X = np.nan_to_num(X, nan=np.nan, posinf=np.nan, neginf=np.nan)

# Impute missing values with mean
imputer = SimpleImputer()
X = imputer.fit_transform(X)

knn = KNeighborsClassifier()

# Split the X and y sets
X_train, X_test, y_train, y_test = train_test_split(X,y,stratify=y,random_state=42)

# Fit knn to the training sets
knn.fit(X_train,y_train)

# Print the score of knn on the test sets
print(knn.score(X_test,y_test))

0.8403361344537815


**Modeling the UFO dataset, part 2**

In [42]:
y = ufo_dropped['type']

In [46]:
from sklearn.naive_bayes import GaussianNB
nb = GaussianNB()

# Use the list of filtered words we created to filter the text vector
filtered_text = desc_tfidf[:, list(filtered_words)]

# Split the X and y sets using train_test_split, setting stratify=y 
X_train, X_test, y_train, y_test = train_test_split(filtered_text.toarray(), y, stratify=y, random_state=42)

# Fit nb to the training sets
nb.fit(X_train,y_train)

# Print the score of nb on the test sets
nb.score(X_test,y_test)

0.13165266106442577